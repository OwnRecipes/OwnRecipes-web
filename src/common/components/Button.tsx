import { forwardRef } from 'react';
import { Button as BootstrapButton } from 'react-bootstrap';
import { Placement } from 'react-bootstrap/esm/types';

import '../css/button.css';

import { CommonProps } from '../types/OverridableComponent';
import ConditionalWrapper from './ConditionalWrapper';
import Tooltip from './Tooltip';

export interface IButtonProps extends CommonProps {
  id:         string;
  type?:      'button' | 'reset' | 'submit';
  variant?:   'primary' | 'secondary' | 'danger' | 'outline-primary' | 'outline-secondary' | 'outline-danger' | 'transparent';
  size?:      'sm' | 'lg';
  disabled?:  boolean;
  onClick?:   (event: React.MouseEvent<HTMLButtonElement>) => void;

  tooltip?:   string;
  tooltipPlacement?: Placement;

  children:   React.ReactNode;
}

const Button = forwardRef<HTMLButtonElement, IButtonProps>(({
  id,
  tooltip, tooltipPlacement,
  children, ...rest }: IButtonProps, ref) => (
    <ConditionalWrapper
        condition = {Boolean(tooltip)}
        render = {childr => <Tooltip id={`${id}-tooltip`} tooltip={tooltip} placement={tooltipPlacement}>{childr}</Tooltip>}>
      <BootstrapButton id={id} aria-label={tooltip || undefined} aria-describedby={undefined} {...rest} ref={ref}>
        {children}
      </BootstrapButton>
    </ConditionalWrapper>
));

export default Button;
