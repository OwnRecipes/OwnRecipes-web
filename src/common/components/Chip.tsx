import { forwardRef } from 'react';
import classNames from 'classnames';

import '../css/chip.css';

import { CommonProps } from '../types/OverridableComponent';

export interface IChipProps extends CommonProps {
  variant: 'primary' | 'secondary';

  children: React.ReactNode;
}

const Chip = forwardRef<HTMLSpanElement, IChipProps>(({
  variant, className, children, ...rest }: IChipProps, ref) => (
    <span
        className={classNames('chip', variant, className)}
        {...rest}
        ref = {ref}>
      {children}
    </span>
));

export default Chip;
