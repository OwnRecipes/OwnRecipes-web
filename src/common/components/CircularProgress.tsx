/** Code adapted from https://github.com/KyleAMathews/react-spinkit */

import { forwardRef } from 'react';
import classNames from 'classnames';

import '../css/circular_progress.css';

import { CommonProps } from '../types/OverridableComponent';

export interface ICircularSpinnerProps extends CommonProps {
  /**
   * Variant of the spinner.
   *
   * @defaultValue `circle`
   */
  variant?: 'circle' | 'three-bounce';
  /**
   * Color of the spinner. When `undefined` the color will be neutral.
   */
  color?: 'primary' | 'secondary';
}

function getDivCount(variant: 'circle' | 'three-bounce'): number {
  switch (variant) {
    case 'circle':       return 12;
    case 'three-bounce': return  3;
    default: throw new Error(`Invalid argument: The spinner variant "${variant}" is invalid.`);
  }
}

const CircularProgress = forwardRef<HTMLDivElement, ICircularSpinnerProps>(({
    variant = 'circle', color, className, ...rest }: ICircularSpinnerProps, ref) => {
  const classes = classNames('circular-progress', variant, color, className);
  const divCount = getDivCount(variant);

  return (
    <div className={classes} {...rest} ref={ref}>
      {/* eslint-disable-next-line react/no-array-index-key */}
      {[...Array(divCount)].map((_, idx) => <div key={idx} />)}
    </div>
  );
});

export default CircularProgress;
