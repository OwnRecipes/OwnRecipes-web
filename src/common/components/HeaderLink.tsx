import { Link } from 'react-router-dom';

import '../css/header_link.css';
import PageScroller from './PageScroller';

export interface IHeaderLinkProps {
  linkFor: string;
}

const HeaderLink: React.FC<IHeaderLinkProps> = ({ linkFor }: IHeaderLinkProps) => (
  <>
    <Link className='headerlink' to={`#${linkFor}`} title='Permalink to this headline'>Â¶</Link>
    <PageScroller uriFragmentId={linkFor} />
  </>
);

export default HeaderLink;
