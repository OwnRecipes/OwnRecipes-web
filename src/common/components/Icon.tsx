import { forwardRef } from 'react';
import classNames from 'classnames';

import '../css/icon.css';

import { CommonProps } from '../types/OverridableComponent';

export interface IIconProps extends CommonProps {
  icon:     string;
  /**
   * @defaultValue `filled`
   */
  variant?: 'filled' | 'light';
  /**
   * @defaultValue `1x`
   */
  size?:    '1x' | '2x';
  /** aria-label as alt for sr. Will be put into a hidden span inside. Use this over aria-label. */
  ariaLabel?: string;
}

function toVariantAffix(variant: 'filled' | 'light'): string {
  return variant === 'filled' ? '-fill' : '';
}

const Icon = forwardRef<HTMLElement, IIconProps>(({
  icon, variant = 'filled', size = '1x',
  ariaLabel,
  className, ...rest }: IIconProps, ref) => (
    <i className={classNames('bi', `bi-${icon}${toVariantAffix(variant)}`, `size-${size}`, className)} {...rest} ref={ref}>
      {ariaLabel && <span className='sr-only'>{ariaLabel}</span>}
    </i>
));

export default Icon;
