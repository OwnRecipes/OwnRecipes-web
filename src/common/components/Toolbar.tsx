import { forwardRef } from 'react';
import classNames from 'classnames';

import '../css/toolbar.css';

import { CommonProps } from '../types/OverridableComponent';

export interface IToolbarProps extends CommonProps {
  position?: 'start' | 'center' | 'end';
  children: React.ReactNode;
}

export const Toolbar = forwardRef<HTMLDivElement, IToolbarProps>(({
  position = 'start',
  children, className, ...rest }: IToolbarProps, ref) => (
    <div className={classNames('toolbar', className, { [position]: position })} {...rest} ref={ref}>
      {children}
    </div>
));
