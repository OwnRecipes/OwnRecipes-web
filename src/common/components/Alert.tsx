import { forwardRef } from 'react';
import { Alert as BootstrapAlert } from 'react-bootstrap';

import '../css/alert.css';

import { CommonProps } from '../types/OverridableComponent';
import Icon from './Icon';
import P from './P';

export interface IAlertProps extends CommonProps {
  severity: 'danger' | 'info';
  title: string;

  children: React.ReactNode;
}

const Alert = forwardRef<HTMLDivElement, IAlertProps>(({ severity, title, children, ...rest }: IAlertProps, ref) => (
  <BootstrapAlert variant={severity} {...rest} ref={ref}>
    <BootstrapAlert.Heading>
      {severity === 'info'   && <Icon icon='info-circle' size='2x' className='alert-icon' />}
      {severity === 'danger' && <Icon icon='exclamation-diamond' size='2x' className='alert-icon' />}
      {title}
    </BootstrapAlert.Heading>
    {typeof children === 'string' && children.length > 0 && (
      <P className='alert-message mb-0'>{children}</P>
    )}
    {typeof children !== 'string' && (
      <div className='alert-message'>{children}</div>
    )}
  </BootstrapAlert>
));

export default Alert;
