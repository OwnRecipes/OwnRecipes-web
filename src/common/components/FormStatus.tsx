import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { defineMessages, useIntl } from 'react-intl';

import '../css/form_errors.css';

import { ValidationError } from '../store/Validation';
import Alert from './Alert';
import NavigationPrompt from './NavigationPrompt';
import P from './P';
import Toast from './Toast';

interface IFormStatusProps {
  dirty: boolean;
  submitting: boolean;
  errors: Record<string, ValidationError> | undefined;

  onSubmitSuccess?: () => void;
}

const FormStatus: React.FC<IFormStatusProps> = ({
    // eslint-disable-next-line arrow-body-style
    dirty, submitting, errors, onSubmitSuccess }: IFormStatusProps) => {
  // console.log(`[FormStatus] dirty=${dirty}, submitting=${submitting}, errors=${JSON.stringify(errors)}`);
  return (
    <>
      <NavigationPrompt isDirty={dirty} />
      <FormErrors errors={errors} />
      <SubmitSuccess dirty={dirty} submitting={submitting} onSubmitSuccess={onSubmitSuccess} errors={errors} />
    </>
  );
};

interface IFormErrorRowErrorProps {
  name: string;
  err: ValidationError;
}

const FormErrorRowError: React.FC<IFormErrorRowErrorProps> = ({ name, err }: IFormErrorRowErrorProps) => {
  const label = document.querySelector(`[data-api-field="${name}"] > label`)?.textContent ?? name;

  const messages: Array<string> = [];
  if (Array.isArray(err)) {
    messages.push(...(err.map(errr => errr.message)));
  } else {
    messages.push(err.message);
  }

  return (
    <tr>
      <td>{label}</td>
      <td>{messages.join('. ')}</td>
    </tr>
  );
};

function getVisibleErrors(errors: Record<string, ValidationError>): Record<string, ValidationError> {
  const visibleErrorsArray = Object.entries(errors).filter(([key]) => !key.startsWith('_'));
  return Object.fromEntries(visibleErrorsArray);
}

interface IFormErrorsProps {
  errors: Record<string, ValidationError> | undefined;
}

const FormErrors: React.FC<IFormErrorsProps> = ({ errors }: IFormErrorsProps) => {
  const { formatMessage } = useIntl();
  const messages = defineMessages({
    form_errors_title: {
      id: 'status.form_errors_title',
      description: 'Title for the form error box.',
      defaultMessage: 'The form contains an error.',
    },
    form_errors_alert: {
      id: 'status.form_errors_alert',
      description: 'Alert text for the form error box.',
      defaultMessage: 'Please fix the described error(s) and continue.',
    },
    form_errors_table_heading_error: {
      id: 'status.form_errors_table_heading_error',
      description: 'The form-errors will be displayed as table, with an error column. This is the column\'s title.',
      defaultMessage: 'Error',
    },
    form_errors_table_heading_message: {
      id: 'status.form_errors_table_heading_message',
      description: 'The form-errors will be displayed as table, with a message column. This is the column\'s title.',
      defaultMessage: 'Message',
    },
  });

  const visibleErrors = useMemo(() => (
    errors != null ? getVisibleErrors(errors) : undefined
  ), [errors]);

  const errorTableRowsJsx = useMemo(() => {
    if (visibleErrors == null || Object.keys(visibleErrors).length === 0) return [];
    const res: Array<React.ReactNode> = [];
    Object.keys(visibleErrors).forEach(key => {
      const nextErr = visibleErrors[key];
      res.push(<FormErrorRowError key={key} name={key} err={nextErr} />);
    });
    return res;
  }, [visibleErrors]);

  if (visibleErrors == null || Object.keys(visibleErrors).length === 0) return null;

  // console.log(`[FormErrors] ${JSON.stringify(visibleErrors)}`);

  const errorTableJsx = (
    <table>
      <thead>
        <tr>
          <th>{formatMessage(messages.form_errors_table_heading_error)}</th>
          <th>{formatMessage(messages.form_errors_table_heading_message)}</th>
        </tr>
      </thead>
      <tbody>
        {errorTableRowsJsx}
      </tbody>
    </table>
  );

  return (
    <Alert
        severity = 'danger'
        className = 'form-errors'
        title = {formatMessage(messages.form_errors_title)}>
      {errorTableJsx}
      <P className='alert-post-message'>{formatMessage(messages.form_errors_alert)}</P>
    </Alert>
  );
};

interface ISubmitSuccessProps {
  dirty: boolean;
  submitting: boolean;
  errors: Record<string, ValidationError> | undefined;

  onSubmitSuccess?: () => void;
}

const SubmitSuccess: React.FC<ISubmitSuccessProps> = ({ dirty, submitting, errors, onSubmitSuccess }: ISubmitSuccessProps) => {
  const { formatMessage } = useIntl();
  const messages = defineMessages({
    save_success: {
      id: 'status.save_success',
      description: 'Toast for successfully saved form.',
      defaultMessage: 'Changes saved.',
    },
  });

  const prevSubmitting = useRef<boolean>(false);
  const [showSaveSuccess, setShowSaveSuccess] = useState<boolean>(false);

  useEffect(() => {
    if (prevSubmitting.current === true && submitting === false && errors == null) {
      setShowSaveSuccess(true);
      onSubmitSuccess?.();
    }

    prevSubmitting.current = submitting;
  }, [submitting]);

  useEffect(() => {
    if (dirty && showSaveSuccess) {
      setShowSaveSuccess(false);
    }
  }, [dirty]);

  const handleCloseSaveSuccessToast = useCallback(() => {
    setShowSaveSuccess(false);
  }, []);

  return (
    <Toast
        show = {showSaveSuccess}
        variant = 'success'
        anchorOrigin={{ horizontal: 'center', vertical: 'bottom' }}
        onClose = {handleCloseSaveSuccessToast}>
      {formatMessage(messages.save_success)}
    </Toast>
  );
};

export default FormStatus;
