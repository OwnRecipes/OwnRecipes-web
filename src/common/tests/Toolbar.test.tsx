import create from '../../test/create';
import { Toolbar } from '../components/Toolbar';

test('Toolbar', () => {
  const component = create(
    <Toolbar>
      <span>x</span>
    </Toolbar>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Toolbar end', () => {
  const component = create(
    <Toolbar position='end'>
      <span>x</span>
    </Toolbar>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
