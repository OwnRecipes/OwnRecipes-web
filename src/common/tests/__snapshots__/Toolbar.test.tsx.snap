// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar 1`] = `
<div
  className="toolbar start"
>
  <span>
    x
  </span>
</div>
`;

exports[`Toolbar end 1`] = `
<div
  className="toolbar end"
>
  <span>
    x
  </span>
</div>
`;
